source/ed/connectors/speed_parser.cpp

```diff
@@ -0,0 +1,172 @@
+    parser.max_speeds_by_type["bridleway"] = walking_speed;
+
+    parser.speed_factor_by_type["motorway"] = 0.9;
+    parser.speed_factor_by_type["motorway_link"] = 0.5;
+    parser.speed_factor_by_type["trunk"] = 0.8;
+    parser.speed_factor_by_type["trunk_link"] = 0.5;
+    parser.speed_factor_by_type["primary"] = 0.8;
+    parser.speed_factor_by_type["secondary"] = 0.6;
+    parser.speed_factor_by_type["tertiary"] = 0.6;
+    parser.speed_factor_by_type["primary_link"] = 0.5;
+    parser.speed_factor_by_type["secondary_link"] = 0.5;
+    parser.speed_factor_by_type["tertiary_link"] = 0.5;
+    parser.speed_factor_by_type["unclassified"] = 0.5;
+    parser.speed_factor_by_type["road"] = 0.5;
+    parser.speed_factor_by_type["residential"] = 0.5;
+    parser.speed_factor_by_type["living_street"] = 0.5;
+    parser.speed_factor_by_type["service"] = 0.2;
+    parser.speed_factor_by_type["track"] = 0.2;
+    parser.speed_factor_by_type["path"] = 0.2;
+    parser.speed_factor_by_type["unclassified"] = 0.5;
+    // useless but prevent warnings
+    parser.speed_factor_by_type["footway"] = 1;
+    parser.speed_factor_by_type["steps"] = 1;
+    parser.speed_factor_by_type["pedestrian"] = 1;
+    parser.speed_factor_by_type["cycleway"] = 1;
+    parser.speed_factor_by_type["bus_stop"] = 1;
+    parser.speed_factor_by_type["platform"] = 1;
+    parser.speed_factor_by_type["construction"] = 1;
+    parser.speed_factor_by_type["bridleway"] = 1;
+
+    return parser;
+}
+
+boost::optional<float> SpeedsParser::get_speed(const std::string& highway,
+                                               const boost::optional<std::string>& max_speed) const {
+    boost::optional<float> speed;
+    boost::optional<float> factor;
+
+    if (max_speeds_by_type.empty() && speed_factor_by_type.empty()) {
+        return boost::none;
+    }
+
+    if (max_speed) {
+        std::string s = boost::algorithm::trim_copy(max_speed.get());
+        auto it = implicit_speeds.find(s);
+        if (it != implicit_speeds.end()) {
+            speed = it->second;
+        } else {
+            try {
+                // todo handle value not in km/h.
+                // if there is no unit the value is in km/h else cast will fail
+                speed = kmh_to_ms(boost::lexical_cast<float>(s));
+            } catch (const std::exception&) {
+                std::cerr << "impossible to convert \"" << max_speed << "\" in speed" << std::endl;
+            }
+        }
+    }
+
+    auto it = max_speeds_by_type.find(highway);
+    if (it != max_speeds_by_type.end()) {
+        update_if_unknown(speed, it->second);
+    } else {
+        std::cerr << highway << " not handled" << std::endl;
+    }
+
+    auto it_factor = speed_factor_by_type.find(highway);
+    if (it_factor != speed_factor_by_type.end() && speed) {
+        speed = speed.get() * it_factor->second;
+    } else {
+        std::cerr << highway << " not handled" << std::endl;
```

Comment:
what will happen if you find the max_speed but not the speed factor? can be thowed an exception?

Reply:
We don't apply any factor, so we will go at `max_speed`, we may want to use a default factor, like `0.5`

Reply:
A default factor of `0.5` would probably be nice :).